kentik_api_library/kentik_api/synthetics/synth_tests/base.py

```diff
@@ -41,55 +44,108 @@ class Defaults:
-            if hasattr(f, "from_dict"):
-                return f.from_dict(v)
+    def from_pb(cls: Type[_ConfigElementT], obj: Any) -> _ConfigElementT:
+        """from_pb can be overridden in a child class to tweak deserialization behavior, and still call _from_protobuf"""
+
+        return cls._from_protobuf(obj)
 
+    @classmethod
+    def _from_protobuf(cls: Type[_ConfigElementT], obj: Any) -> _ConfigElementT:
+        def get_value(dst_type: Type[Any], src_value: Any) -> Any:
+            if hasattr(dst_type, "from_pb"):
+                return dst_type.from_pb(src_value)
             try:
-                return f(v)
+                return dst_type(src_value)
             except TypeError as error:
-                if f._name == "List":
-                    return [get_value(type(i), i) for i in v]
-                if f._name == "Dict":
-                    return {_k: get_value(type(_v), _v) for _k, _v in v.items()}
-                raise RuntimeError(f"Don't know how to instantiate '{f}' (value: '{v}')") from error
-
-        _init_fields = {f.name: f for f in fields(cls) if f.init}
-        args = {k: get_value(_init_fields[k].type, v) for k, v in d.items() if k in _init_fields.keys()}
-        # noinspection PyArgumentList
-        inst = cls(**args)  # type: ignore
-        _other_fields = {f.name: f for f in fields(cls) if not f.init}
-        for n, f in _other_fields.items():
-            if n[0] == "_":
-                k = n.split("_")[1]
-            else:
-                k = n
-            if k in d:
-                setattr(inst, n, get_value(f.type, d[k]))
-        return inst
+                if dst_type._name == "List":
+                    return [get_value(get_args(dst_type)[0], i) for i in src_value]
+                if dst_type._name == "Dict":
+                    return {_k: get_value(get_args(dst_type)[1], _v) for _k, _v in src_value.items()}
+                raise RuntimeError(f"Don't know how to instantiate '{dst_type}' (value: '{src_value}')") from error
+
+        _init_fields = [f for f in fields(cls) if f.init]
+        args = {}
+        for f in _init_fields:
+            src_name = f.name[1:] if (f.name[0] == "_") else f.name  # handle read-only fields that start with "_"
+            args[f.name] = get_value(f.type, getattr(obj, src_name))
+        instance = cls(**args)
+
+        _remaining_fields = [f for f in fields(cls) if not f.init]
+        for f in _remaining_fields:
+            src_name = f.name[1:] if (f.name[0] == "_") else f.name  # handle read-only fields that start with "_"
+            v = get_value(f.type, getattr(obj, src_name))
+            setattr(instance, f.name, v)
+        return instance
+
+
+DateTimeT = TypeVar("DateTimeT", bound="DateTime")
+
+
+class DateTime(datetime):
+    """
+    A datetime type with protobuf serialization logic
+    """
+
+    PB_TYPE = DoNotSerializeMarker  # datetime fields are read-only - provided by the server. Skip serialization
```

Comment:
Ah, I've missed this the last time: Unfortunately, `test.edate` may be passed in `TestUpdate` requests in order to guard against modifying data that has been modified since the last retrieval. It is not required, but if it is set, the API returns error if the `edate` in the request is older than `edate` of the test in the DB.

Reply:
Ok, so edate  for Test is read-write. Updating

Reply:
Well, sort of. It is ignored in all requests except for `TestUpdate`.